src/mapping/query/arithmetic.rs

```diff
@@ -0,0 +1,406 @@
+use super::Function;
+use crate::{
+    event::{Event, Value},
+    mapping::Result,
+};
+use bytes::BytesMut;
+
+#[derive(Debug, Clone)]
+pub(in crate::mapping) enum Operator {
+    Multiply,
+    Divide,
+    Modulo,
+    Add,
+    Subtract,
+    Equal,
+    NotEqual,
+    Greater,
+    GreaterOrEqual,
+    Less,
+    LessOrEqual,
+    And,
+    Or,
+}
+
+#[derive(Debug)]
+pub(in crate::mapping) struct Arithmetic {
+    left: Box<dyn Function>,
+    right: Box<dyn Function>,
+    op: Operator,
+}
+
+impl Arithmetic {
+    pub(in crate::mapping) fn new(
+        left: Box<dyn Function>,
+        right: Box<dyn Function>,
+        op: Operator,
+    ) -> Self {
+        Self { left, right, op }
+    }
+}
+
+/// If either value is a floating point number type and the other value is an
+/// integer type the integer is "degraded" into a float. This allows us to
+/// perform arithmetic on common values, but if both are integers then their
+/// precision is preserved.
+fn consistent_number_types(mut left: Value, mut right: Value) -> (Value, Value) {
+    if let Value::Float(_) = left {
+        if let Value::Integer(ri) = right {
+            right = Value::Float(ri as f64);
+        }
+    } else if let Value::Float(_) = right {
+        if let Value::Integer(li) = left {
+            left = Value::Float(li as f64);
+        }
+    }
+    (left, right)
+}
+
+// Degrades non-float numerical types into floats for the purposes of convenient
+// boolean comparison.
+fn compare_number_types(
+    left: Value,
+    right: Value,
+    compare_fn: &dyn Fn(f64, f64) -> bool,
+) -> Result<Value> {
+    let (left, right) = consistent_number_types(left, right);
+    Ok(Value::Boolean(match left {
+        Value::Float(fl) => match right {
+            Value::Float(fr) => compare_fn(fl, fr),
+            vr => return Err(format!("unable to compare right-hand field type {:?}", vr)),
+        },
+        Value::Integer(il) => match right {
+            Value::Integer(ir) => compare_fn(il as f64, ir as f64),
+            vr => return Err(format!("unable to compare right-hand field type {:?}", vr)),
+        },
+        vl => return Err(format!("unable to compare left-hand field type {:?}", vl)),
+    }))
```

Comment:
See above for a pattern using `match (left, right)` for much more compact logic, or even just `match consistent_number_types(left, right)`. Ditto for all the pair matches in `fn execute` below.